rust/automerge-c/src/doc/mark.rs

```diff
@@ -0,0 +1,294 @@
+use am::ReadDoc;
+
+use crate::{
+    byte_span::{to_str, AMbyteSpan},
+    item::{AMitem, Item},
+    items::AMitems,
+    obj::{to_obj_id, AMobjId},
+    result::{to_result, AMresult},
+};
+
+use super::{
+    utils::{to_doc, to_doc_mut},
+    AMdoc,
+};
+
+macro_rules! to_expand_mark {
+    ($mark_expand:expr) => {{
+        let result: Result<ExpandMark, am::AutomergeError> = (&$mark_expand).try_into();
+        match result {
+            Ok(expand_mark) => expand_mark,
+            Err(e) => return AMresult::error(&e.to_string()).into(),
+        }
+    }};
+}
+
+pub(crate) use to_expand_mark;
+
+/// \ingroup enumerations
+/// \enum AMmarkExpand
+/// \installed_headerfile
+/// \brief A mark's expansion mode for when bordering text is inserted.
+#[derive(PartialEq, Eq)]
+#[repr(u8)]
+pub enum AMmarkExpand {
+    /// Include text inserted at the end offset.
+    After = 3,
+    /// Include text inserted at the start offset.
+    Before = 2,
+    /// Include text inserted at either offset.
+    Both = 4,
+    /// The default tag, not a mark expansion mode signifier.
+    Default = 0,
+    /// Exclude text inserted at either offset.
+    None = 1,
+}
+
+impl Default for AMmarkExpand {
+    fn default() -> Self {
+        Self::Default
+    }
+}
+
+impl TryFrom<&AMmarkExpand> for ExpandMark {
+    type Error = am::AutomergeError;
+
+    fn try_from(mark_expand: &AMmarkExpand) -> Result<Self, Self::Error> {
+        use am::AutomergeError::InvalidValueType;
+        use AMmarkExpand::*;
+
+        match mark_expand {
+            After => Ok(Self::After),
+            Before => Ok(Self::Before),
+            Both => Ok(Self::Both),
+            None => Ok(Self::None),
+            _ => Err(InvalidValueType {
+                expected: type_name::<Self>().to_string(),
+                unexpected: type_name::<u8>().to_string(),
+            }),
+        }
+    }
+}
+
+pub struct AMmark<'a>(Mark<'a>);
+
+impl<'a> AMmark<'a> {
+    pub fn new(mark: Mark<'a>) -> Self {
+        Self(mark)
+    }
+}
+
+impl<'a> AsRef<Mark<'a>> for AMmark<'a> {
+    fn as_ref(&self) -> &Mark<'a> {
+        &self.0
+    }
+}
+
+/// \memberof AMmark
+/// \brief Gets the name of a mark.
+///
+/// \param[in] mark A pointer to an `AMmark` struct.
+/// \return A UTF-8 string view as an `AMbyteSpan` struct.
+/// \pre \p mark `!= NULL`
+/// \post `(`\p mark `== NULL) -> (AMbyteSpan){NULL, 0}`
+/// \internal
+///
+/// # Safety
+/// mark must be a valid pointer to an AMmark
+#[no_mangle]
+pub unsafe extern "C" fn AMmarkName(mark: *const AMmark) -> AMbyteSpan {
+    if let Some(mark) = mark.as_ref() {
+        if let Ok(name) = mark.as_ref().name().as_bytes().try_into() {
+            return name;
+        }
+    }
+    Default::default()
+}
+
+/// \memberof AMmark
+/// \brief Gets the value of a mark.
+///
+/// \param[in] mark A pointer to an `AMmark` struct.
+/// \return A pointer to an `AMresult` struct with an `AMitem` struct
+///          containing the value.
```

Comment:
@ConradIrwin, every other function that returns an unpredictable type of value is documented as
```suggestion
/// \return A pointer to an `AMresult` struct with an `AMitem` struct.
```